### PR TITLE
[20166] Fix data race in PDPListener and SecurityManager

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

This PR fixes a data race occurring in `PDPListenerSecurityInitiator`  affecting `SecurityManager`. The `temp_participant_data_` was being overwritten as `discovered_participant()` was being called without the `PDP` mutex taken. This causes a corruption when the secure participant was in the middle of matching to a previous secure participant.

This PR also ensures that the `discovered_participant()` method is using the `ParticipantProxyData` from the `discovered_participants` collection.

_Note: openning as draft as proper test needs to be added_

@Mergifyio backport 2.12.x 2.11.x 2.10.x

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [X] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/security/SecurityManager.cpp [hunk 1]
```diff
@@ -589,6 +589,8 @@ bool SecurityManager::discovered_participant(
     // Create or find information
     bool undiscovered = false;
     DiscoveredParticipantInfo::AuthUniquePtr remote_participant_info;
+    //! Use the information from the collection
```
Comment (MiguelCompany) [inline src/cpp/rtps/security/SecurityManager.cpp:592]:
```suggestion
    // Use the information from the collection
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Apart from my NIT below, a regression test should be added

**Comment by Mario-DL** [comment]
@richiprosima please test this

File: src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.cpp [hunk 2]
```diff
@@ -71,25 +71,19 @@ void PDPSecurityInitiatorListener::process_alive_data(
...
-            // Unlock the mutexes before calling the callback
-            reader->getMutex().unlock();
-            lock.unlock();
-
-            response_cb_(callback_data);
-
-            // Take the locks back
-            lock.lock();
-            reader->getMutex().lock();
+            response_cb_(temp_participant_data_);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.cpp:82]:
```suggestion
        auto callback_data = new_data;
        reader->getMutex().unlock();
        lock.unlock();

        //! notify security manager in order to start handshake
        bool ret = parent_pdp_->getRTPSParticipant()->security_manager().discovered_participant(callback_data);
        //! Reply to the remote participant
        if (ret)
        {
            response_cb_(callback_data);
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.cpp (hunk 2 above)
Comment (Mario-DL) [reply src/cpp/rtps/builtin/discovery/participant/DS/PDPSecurityInitiatorListener.cpp]: I really like the suggestion

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Comment by Mario-DL** [comment]
@richiprosima please test this

File: test/blackbox/common/BlackboxTestsSecurity.cpp [hunk 3]
```diff
@@ -3848,6 +3848,95 @@ TEST(Security, AllowUnauthenticatedParticipants_TwoParticipantsDifferentCertific
     ASSERT_FALSE(writer.is_matched());
 }
 
+// Regresion test for Refs #20166
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsSecurity.cpp:3851]:
```suggestion
// Regresion test for redmine issue 20166
```

File: test/blackbox/common/BlackboxTestsSecurity.cpp [hunk 4]
```diff
@@ -3848,6 +3848,95 @@ TEST(Security, AllowUnauthenticatedParticipants_TwoParticipantsDifferentCertific
...
+    pub_property_policy.properties().emplace_back(Property("dds.sec.crypto.plugin",
+            "builtin.AES-GCM-GMAC"));
+    pub_property_policy.properties().emplace_back(Property("dds.sec.access.plugin",
+            "builtin.Access-Permissions"));
+    pub_property_policy.properties().emplace_back(Property("dds.sec.access.builtin.Access-Permissions.permissions_ca",
+            "file://" + std::string(certs_path) + "/maincacert.pem"));
+    pub_property_policy.properties().emplace_back(Property("dds.sec.access.builtin.Access-Permissions.governance",
+            "file://" + std::string(certs_path) + "/governance_helloworld_all_enable.smime"));
+    pub_property_policy.properties().emplace_back(Property("dds.sec.access.builtin.Access-Permissions.permissions",
+            "file://" + std::string(certs_path) + "/permissions_helloworld.smime"));
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsSecurity.cpp:3883]: Would be nice if you could do something similar to [these lines](https://github.com/eProsima/Fast-DDS/blob/ef14749e348fac14555e280441ca7abe6bf48af5/test/blackbox/common/BlackboxTestsSecurity.cpp#L4808-L4811)

File: test/blackbox/common/BlackboxTestsSecurity.cpp [hunk 5]
```diff
@@ -3848,6 +3848,95 @@ TEST(Security, AllowUnauthenticatedParticipants_TwoParticipantsDifferentCertific
...
+TEST(Security, InANonSecureParticipantWithTwoSecureParticipantScenario_TheTwoSecureParticipantsCorrectlyCommunicate)
+{
+    //! Create
+    PubSubReader<HelloWorldPubSubType> non_secure_reader("HelloWorldTopic");
+    PubSubReader<HelloWorldPubSubType> secure_reader("HelloWorldTopic");
+    PubSubWriter<HelloWorldPubSubType> secure_writer("HelloWorldTopic");
+
+    non_secure_reader.history_depth(10).
+            reliability(eprosima::fastrtps::RELIABLE_RELIABILITY_QOS).
+            // Force a high number of announcements. Publish 1 every 1 ms
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsSecurity.cpp:3861]: We shouldn't need this if we follow the initialization order that was reported:

1. secure_writer
2. non_secure_reader
3. secure_reader

File: test/blackbox/common/BlackboxTestsSecurity.cpp [hunk 6]
```diff
@@ -3848,6 +3848,95 @@ TEST(Security, AllowUnauthenticatedParticipants_TwoParticipantsDifferentCertific
     ASSERT_FALSE(writer.is_matched());
 }
 
+// Regresion test for Refs #20166
+TEST(Security, InANonSecureParticipantWithTwoSecureParticipantScenario_TheTwoSecureParticipantsCorrectlyCommunicate)
+{
+    //! Create
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsSecurity.cpp:3854]: Please remove the `!` from all the comments

File: test/blackbox/common/BlackboxTestsSecurity.cpp [hunk 7]
```diff
@@ -3848,6 +3848,95 @@ TEST(Security, AllowUnauthenticatedParticipants_TwoParticipantsDifferentCertific
...
+    auto data = default_helloworld_data_generator();
+
+    secure_reader.startReception(data);
+
+    secure_writer.send(data);
+
+    // In this test all data should be sent.
+    ASSERT_TRUE(data.empty());
+
+    secure_reader.block_for_all();
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsSecurity.cpp:3937]: I would take the chance to ensure the non_secure_reader does not receive data:

```suggestion
    secure_reader.startReception(data);
    non_secure_reader.startReception(data);

    secure_writer.send(data);

    // In this test all data should be sent.
    ASSERT_TRUE(data.empty());

    secure_reader.block_for_all();
    EXPECT_EQ(non_secure_reader.getReceivedCount(), 0u);
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Pull request merged**